src/java.base/share/classes/sun/security/provider/HSS.java

```diff
@@ -0,0 +1,834 @@
+                for (int i = 0; i < p; i++) {
+                    int a = coef(QWithChecksum, i);
+                    if (a == twoPowWMinus1) {
+                        lmOtSig.getY(i, preCandidate, 22 + i * n);
+                    } else {
+                        preZi[20] = (byte) (i >> 8);
+                        preZi[21] = (byte) i;
+                        lmOtSig.getY(i, preZi, 23);
+                    }
+
+                    for (int j = a; j < twoPowWMinus1; j++) {
+                        preZi[22] = (byte) j;
+                        if (j < twoPowWMinus2) {
+                            digestFixedLengthPreprocessed(
+                                    sha256, preZi, hashLen, preZi, 23, n);
+                        } else {
+                            digestFixedLengthPreprocessed(
+                                    sha256, preZi, hashLen, preCandidate, 22 + i * n, n);
+                        }
+                    }
+                }
+                md.update(preCandidate, 0, 22 + p * n);
+
+                result = md.digest();
+            } catch (DigestException e) { // This should not happen
+                throw new ProviderException("Digest failed", e);
+            }
+            if (n != hashAlg_n) {
+                result = Arrays.copyOfRange(result, 0, n);
+            }
+            return result;
+        }
+    }
+
+    public static class KeyFactoryImpl extends KeyFactorySpi {
+        @Override
+        protected PublicKey engineGeneratePublic(KeySpec keySpec)
+                throws InvalidKeySpecException {
+            if (keySpec instanceof X509EncodedKeySpec) {
+                try {
+                    X509EncodedKeySpec x509Spec = (X509EncodedKeySpec)keySpec;
+                    return new HSSPublicKey(
+                            x509Spec.getEncoded(), true);
+                } catch (InvalidKeyException e) {
+                    throw new InvalidKeySpecException(e);
+                }
+            } else if (keySpec instanceof RawKeySpec x) {
+                try {
+                    return new HSSPublicKey(x.getKeyArr(), false);
+                } catch (InvalidKeyException e) {
+                    throw new InvalidKeySpecException(e);
+                }
+            }
+            throw new InvalidKeySpecException("Unrecognized KeySpec");
+        }
+
+        @Override
+        protected PrivateKey engineGeneratePrivate(KeySpec keySpec)
+                throws InvalidKeySpecException {
+            throw new InvalidKeySpecException(
+                    "Private key generation is not supported");
+        }
+
+        @Override
+        protected <T extends KeySpec> T engineGetKeySpec(Key key, Class<T> keySpec)
+                throws InvalidKeySpecException {
+            if (key == null) {
+                throw new InvalidKeySpecException("key should not be null");
+            }
+            if (key.getFormat().equals("X.509") &&
```

Comment:
Would it make sense to also check if `key` is an instance of `HSSPublicKey`?

Reply:
I don't think so, the check here is not too heavy, it covers that case and it is called rarely enough that this kind of optimisation would be worthwhile